joblib/my_exceptions.py

```diff
@@ -64,10 +62,10 @@ def _mk_exception(exception, name=None):
             # We cannot create a subclass: we are already a trivial
             # subclass
             return JoblibException, this_name
-        this_exception = type(this_name, (exception, JoblibException),
-                    dict(__repr__=JoblibException.__repr__,
-                        __str__=JoblibException.__str__),
-                    )
+        elif issubclass(exception, JoblibException):
+            return JoblibException, JoblibException.__name__
+        this_exception = type(
+            this_name, (JoblibException, exception), {})
```

Comment:
This is the main fix: I changed the ordering of the bases.

I am not sure the reason why `exception` was before `JoblibException` previously. Amongst other things, this means that we don't control over the `__init__` of the generated exception. AFAICS the only reason we want the generated exception to be a subclass of `exception` is that we can catch exceptions with the same name, irrespective whether the computation happens in parallel or not. If that's the only reason, I don't see why we can not put `exception` as the last base.

@ogrisel @GaelVaroquaux any thoughts?